### PR TITLE
Introduce stack definition macros

### PR DESCRIPTION
**Pull request by tyrantlizard77**
We need a good method for defining stack memory regions. Right now the typical way to do it is with __stack __noinit. We need something more generic so that we can have fine, architecture-specific control on how the stacks are declared.

This patch series replaces all instances of __stack with K_THREAD_STACK_DECLARE() and related macros.

It is expected that arches will implement their own versions of these macros to implement whatever behind-the-scenes magic is necessary to set up these stacks for MMU or MPU-based memory protection.

### CODE REVIEW

**Review by galak — CHANGES_REQUESTED** [review]
Can you give an example of use of __DEPRECATED_MACRO in the commit message.

**Comment by andrewboie** [comment]
Whoops, git ate the line that started with #define in the commit message. Should be fixed now.

File: tests/kernel/' [hunk 1]
```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2016 Intel Corporation
```
Comment (andrewboie) [inline tests/kernel/']: This should be deleted, mis-type in my text editor

File: include/toolchain/common.h [hunk 2]
```diff
@@ -98,7 +98,8 @@
 /* Indicate that an array will be used for stack space. */
 
 #if !defined(_ASMLANGUAGE)
-  #define __stack __aligned(STACK_ALIGN)
+  /* don't use this anymore, use K_DECLARE_STACK instead. Remove for 1.11 */
```
Comment (lpereira) [inline include/toolchain/common.h]: Do we have a macro for version checking? Might be useful for future deprecations:

```C
#if ZEPHYR_CHECK_VERSION(1,11)
#error Remove definition for __stack
#endif
``

(glib/gtk has a similar macro, and it's pretty useful.)

File: include/toolchain/gcc.h [hunk 3]
```diff
@@ -94,6 +94,11 @@ do {                                                                    \
 #define __weak __attribute__((__weak__))
 #define __unused __attribute__((__unused__))
 
+/* Be *very* careful with this, you cannot filter out with -wno-deprecated,
+ * which has implications for -Werror
+ */
+#define __DEPRECATED_MACRO _Pragma("GCC warning \"Macro is deprecated\"")
```
Comment (lpereira) [inline include/toolchain/gcc.h]: The suggested macro can even be used like so:

```C
#define __DEPRECATED_MACRO(cond)  __DEPRECATED_MACRO_##cond
#define __DEPRECATED_MACRO_0 _Pragma("GCC warning \"Macro is deprecated\"")
#define __DEPRECATED_MACRO_1 _Pragma("GCC error \"Gardening required on this macro\"")

#define __DEPRECATE_MACRO_ON(maj, min) \
   __DEPRECATED_MACRO(ZEPHYR_VERSION_CHECK(maj,min))
```

And we can then use `__DEPRECATE_MACRO_ON(1,11)`, ensuring a compile error will happen. (`ZEPHYR_VERSION_CHECK()` should evaluate to `0` or `1` given maj/min).

File: include/toolchain/common.h (hunk 2 above)
Comment (nashif) [inline include/toolchain/common.h]: yes we do, similar to the linux kernel:

#if ZEPHYR_VERSION_CODE <= ZEPHYR_VERSION(1,8,0)
foo
#else 
bar
#endif

**Review by jukkar — COMMENTED** [review]
LGTM

File: include/toolchain/gcc.h (hunk 3 above)
Comment (andrewboie) [inline include/toolchain/gcc.h]: What you're proposing here isn't possible with the C preprocessor, you can't implement ZEPHYR_VERSION_CHECK(maj, min) in this way.
Best I can do is a #if ... #else on the SW version at the __stack definition I am trying to deprecate.

File: include/toolchain/common.h (hunk 2 above)
Comment (andrewboie) [inline include/toolchain/common.h]: Anas, this doesn't work either, the preprocessor does not expand ZEPHYR_VERSION the way you expect:

/projects/zephyr2/include/toolchain/common.h:102:43: error: missing binary operator before token "("
 #if (ZEPHYR_VERSION_CODE <= ZEPHYR_VERSION(1,11,0))
                                           ^
/projects/zephyr2/include/toolchain/common.h:100:0: error: unterminated #if
 #if !defined(_ASMLANGUAGE)
 
Also, as noted below, what Leandro is proposing is not actually possible with the preprocessor.

I ask we leave this macro as-is and focus on the memory protection features in this patch set.

Comment (nashif) [inline include/toolchain/common.h]: You need to include version.h if I am not mistaken.

**Review by cvinayak — APPROVED** [review]
Stack related macros look good to me.

**Review by nashif — APPROVED** [review]

**Pull request merged**